packages/dht/src/dht/routing/RoutingSession.ts

```diff
@@ -0,0 +1,222 @@
+import { PeerDescriptor } from '../../exports'
+import { DhtPeer } from '../DhtPeer'
+import { SortedContactList } from '../contact/SortedContactList'
+import { PeerID, PeerIDKey } from '../../helpers/PeerID'
+import { Logger } from '@streamr/utils'
+import EventEmitter from 'eventemitter3'
+import { v4 } from 'uuid'
+import { RouteMessageWrapper } from '../../proto/packages/dht/protos/DhtRpc'
+import { RemoteRouter } from './RemoteRouter'
+import { RoutingRpcCommunicator } from '../../transport/RoutingRpcCommunicator'
+import { RoutingServiceClient } from '../../proto/packages/dht/protos/DhtRpc.client'
+import { toProtoRpcClient } from '@streamr/proto-rpc'
+
+const logger = new Logger(module)
+
+const MAX_FAILED_HOPS = 2
+
+export interface RoutingSessionEvents {
+    // This event is emitted when a peer responds with a success ack
+    // to routeMessage call
+    routingSucceeded: (sessionId: string) => void
+    partialSuccess: (sessionId: string) => void
+
+    // This event is emitted when all the candidates have been gone
+    // through, and none of them responds with a success ack
+    routingFailed: (sessionId: string) => void
+    stopped: (sessionId: string) => void
+    noCandidatesFound: (sessionId: string) => void
+}
+
+export enum RoutingMode { ROUTE, FORWARD, RECURSIVE_FIND }
+
+export class RoutingSession extends EventEmitter<RoutingSessionEvents> {
+
+    public readonly sessionId = v4()
+    private readonly rpcCommunicator: RoutingRpcCommunicator
+    private ongoingRequests: Set<PeerIDKey> = new Set()
+    private contactList: SortedContactList<RemoteRouter>
+    private readonly ownPeerDescriptor: PeerDescriptor
+    private readonly messageToRoute: RouteMessageWrapper
+    private connections: Map<PeerIDKey, DhtPeer>
+    private readonly parallelism: number
+    private failedHopCounter = 0
+    private successfulHopCounter = 0
+    private readonly mode: RoutingMode = RoutingMode.ROUTE
```

Comment:
Value always set in constructor

```suggestion
    private readonly mode: RoutingMode
```